cli: Improve error message on multiwallet cli-side commands

pablomartin4btc opened the pull request:
Running a CLI command when multiple wallets are loaded and `-rpcwallet` is not specified, should return a clearer error.

Currently in `master`:

```
$ bitcoin-cli -regtest -generate 1
error code: -19
error message:
Wallet file not specified (must request wallet RPC through /wallet/<filename> uri-path).
Try adding "-rpcwallet=<filename>" option to bitcoin-cli command line.
```

With this change:

```
$ bitcoin-cli -regtest -generate 1
error code: -19
error message:
Multiple wallets are loaded. Please select which wallet to use by requesting the RPC through the /wallet/<walletname> URI path. Or for the CLI, specify the "-rpcwallet=<walletname>" option before the command (run "bitcoin-cli -h" for help or "bitcoin-cli listwallets" to see which wallets are currently loaded).
```

pablomartin4btc committed:
rpc, cli: improve error message on multiwallet mode

The primary objective is to provide users with clearer
and more informative error messages when encountering
the RPC_WALLET_NOT_SPECIFIED error, which occurs when
multiple wallets are loadad.

This commit also rectifies the error message consistency
by bringing the error message in line with the definition
established in protocol.h ("error when there are multiple
wallets loaded").

kouloumos reviewed (commented):
The behavior that you are describing is no different than running any wallet RPC with multiple loaded wallets but without adding the `-rpcwallet=` option.  
e.g. `bitcoin-cli getnewaddress -rpcwallet=testWallet1` will fail because `[options]` should precede the `command`.

I understand the confusion but I believe that its source is not `-generate` but our definition of `[options]`. The help text specifies usage of the cli as:
```
bitcoin-cli [options] <command> [params] Send command to Bitcoin Core
```

Usually `[options]` refer to optional parameters or flags that allow the user to customize the execution of the command to either modify its default behavior or provide additional information. 

And that was true until client-side commands became part of `[options]`:
- First,`-getinfo` was introduced with #10871 as a way to offer client-side the capabilities of the now removed `getinfo` RPC ([history of getinfo removal](https://github.com/bitcoin/bitcoin/pull/13891#issuecomment-412835513)).
- Then `-generate` was added with #19133 based on the same reasoning, as a way to offer a client-side version of the [deprecated generate RPC](https://github.com/bitcoin/bitcoin/issues/14299).
- And then followed by the additions of [`-netinfo`](https://github.com/bitcoin/bitcoin/pull/19643) and [`-addinfo`](https://github.com/bitcoin/bitcoin/pull/21595)

The order of `[options]` shouldn't matter, but right now those client-side commands are part of options and in the case of `-generate` (which accepts arguments) the order matters, therefore it creates confusion.

I think that a better option could be to decoupled those from `[options]` . It doesn't seem right that `bitcoin-cli -generate -netinfo -addrinfo -getinfo` is a valid command.

pablomartin4btc commented:
Thanks @kouloumos for taking the time to review this PR, I agree with your comments and I appreciate you added more context into it so people could bring other thoughts.

I just wanted to make a small change first and specifically to the `-generate` code block but since, as you said, the confusion applies to any other "client-side command" or `[options]` I could move the validation a bit up in the code around the line [where the args variable is set](https://github.com/bitcoin/bitcoin/blob/b5868f4b1f884e8d6612f34ca4005fe3a992053d/src/bitcoin-cli.cpp#L1122), which contains  the [params] of the command (or the args for -generate).

I got another commit that I could publish that changes ArgsManager in order to accept, for example, `-rpcwallet` as an independent command or `[option]` after `-generate` so `bitcoin-cli -generate 3 4 -rpcwallet=X` works (so order won't matter anymore) and I got something similar to the validation you mention in your last sentence.

hernanmarino reviewed (commented):
tested ACK. Changes provide more feedback to the user on error messages. Works as intended.

jonatack reviewed (commented):
@pablomartin4btc as mentioned in https://github.com/bitcoin/bitcoin/pull/26990#pullrequestreview-1274911604, I think a general solution would be more helpful regarding the CLI/RPC error messages, for the case when multiple wallets are loaded and the user needs to specify which wallet to use, while also clarifying that the -rpcwallet option needs to be before the CLI command.

What do you think about an approach like this: https://github.com/jonatack/bitcoin/commits/2023-02-improve-rpcwallet-error-messages

hernanmarino reviewed (commented):
Light code review ACK 86c0442da27ed92855f9722eb10d94e72b197f2d Also tested ACK the use cases refered by author [here](https://github.com/bitcoin/bitcoin/pull/26990#issuecomment-1446919139) 
The only thing i suggest changing is giving a more descriptive error on the following situation :

```
$ bitcoin-cli -regtest -generate 1 -rpcwallet=tst2
error: Error: this is not a valid cli-command argument "-rpcwallet=tst2", please check the syntax of the "-generate" command.
```
I would only show that error when the extra / invalid parameter does not start with "-" . When starting with "-" , perhaps an error similar to the following might be more useful :
```
$ bitcoin-cli -regtest -generate 1 -rpcwallet=tst2
error: Error: this is not a valid cli-command argument "-rpcwallet=tst2", if that is a valid option you might try passing it before -generate
```
This is just a sample text for the error, it might apply to other options requiring parameters and not only to -generate.

pablomartin4btc commented:
> The only thing i suggest changing is giving a more descriptive error on the following situation :
> ```
> $ bitcoin-cli -regtest -generate 1 -rpcwallet=tst2
> error: Error: this is not a valid cli-command argument "-rpcwallet=tst2", if that is a valid option you might try passing it before -generate
> ```

Thanks, I agree with it, updated it.

hernanmarino reviewed (approved):
re ACK 7f5fd1e6999b955dbb0b382bbfdce90a02e1cd8e Looks good to me now

andrewtoth commented on src/bitcoin-cli.cpp:
```diff
@@ -887,7 +887,8 @@ static void ParseError(const UniValue& error, std::string& strPrint, int& nRet)
         }
         if (err_code.isNum() && err_code.getInt<int>() == RPC_WALLET_NOT_SPECIFIED) {
-            strPrint += "\nTry adding \"-rpcwallet=<filename>\" option to bitcoin-cli command line.";
+            strPrint += " For the CLI, specify the \"-rpcwallet=<filename>\" option before the command";
```
I'm not sure if this new message makes it more clear. Specifically, "For the CLI" seems like it might confuse users because there's nothing opposed to that in the message.

pablomartin4btc replied:
I've updated the test with the full error message, also I've changed the text a bit:
 
`Multiple wallets are loaded. Please specify a wallet by requesting the RPC through the /wallet/<filename> URI path. Using "bitcoin-cli", add the "-rpcwallet=<filename>" option before the command (run "bitcoin-cli -h" for help or "bitcoin-cli listwallets" to see which wallets are currently loaded).`

pablomartin4btc replied:
I've re-tweaked it a bit, added a `\n` as current implementation and just set the error message that comes from the `utils` in the test's constant:
```
Multiple wallets are loaded. Please specify a wallet to use by requesting the RPC through the /wallet/<filename> URI path.
Using "bitcoin-cli", add the "-rpcwallet=<filename>" option before the command.
Run "bitcoin-cli -h" for help or "bitcoin-cli listwallets" to see which wallets are currently loaded.
````
WALLET_NOT_SPECIFIED = 'Multiple wallets are loaded. Please select a wallet to use'`

pablomartin4btc replied:
I've re-considered your [preference](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1614026697) [above](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1127282385) @jonatack.

andrewtoth commented on src/bitcoin-cli.cpp:
```diff
@@ -887,7 +887,8 @@ static void ParseError(const UniValue& error, std::string& strPrint, int& nRet)
         if (err_code.isNum() && err_code.getInt<int>() == RPC_WALLET_NOT_SPECIFIED) {
-            strPrint += "\nTry adding \"-rpcwallet=<filename>\" option to bitcoin-cli command line.";
+            strPrint += " For the CLI, specify the \"-rpcwallet=<filename>\" option before the command";
+            strPrint += " (run \"bitcoin-cli -h\" for help or \"bitcoin-cli listwallets\" to see which wallets are currently loaded).";
```
Where does `bitcoin-cli -h` show how to solve this problem specifically? Telling users to run that here might not be very helpful. The second part could be useful information though.

pablomartin4btc replied:
@andrewtoth I've updated the error message, check if that looks better for you.

andrewtoth commented on src/bitcoin-cli.cpp:
```diff
@@ -1078,6 +1079,68 @@ static UniValue GetNewAddress()
+ */
+bool IsExclusivelyCliCommand(const std::string cliCommand)
+{
+    // Initialize array with cli-commands that run exclusively
```
nit: `Initialize set`

pablomartin4btc replied:
fixed

andrewtoth commented on src/wallet/rpc/util.cpp:
```diff
@@ -89,7 +89,7 @@ std::shared_ptr<CWallet> GetWalletForJSONRPCRequest(const JSONRPCRequest& reques
     }
     throw JSONRPCError(RPC_WALLET_NOT_SPECIFIED,
-        "Wallet file not specified (must request wallet RPC through /wallet/<filename> uri-path).");
+        "Multiple wallets are loaded. Please select which wallet to use by requesting the RPC through the /wallet/<filename> URI path.");
```
```suggestion
        "Multiple wallets are loaded. Specify the wallet by requesting the RPC through /wallet/<filename> URI path.");
```

andrewtoth commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -28,7 +28,10 @@
 TOO_MANY_ARGS = 'error: too many arguments (maximum 2 for nblocks and maxtries)'
 WALLET_NOT_LOADED = 'Requested wallet does not exist or is not loaded'
-WALLET_NOT_SPECIFIED = 'Wallet file not specified'
+WALLET_NOT_SPECIFIED = 'For the CLI, specify the "-rpcwallet=<filename>" option before the command'
```
Again, I don't think "For the CLI" message is useful if it doesn't have a "as opposed to the ".

pablomartin4btc replied:
I've updated the error message as I've commented it [here](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1131669863).

andrewtoth commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -28,7 +28,10 @@
 WALLET_NOT_LOADED = 'Requested wallet does not exist or is not loaded'
-WALLET_NOT_SPECIFIED = 'Wallet file not specified'
+WALLET_NOT_SPECIFIED = 'For the CLI, specify the "-rpcwallet=<filename>" option before the command'
+INVALID_CLI_COMMAND_ARGUMENT = 'Error: this is not a valid cli-command argument \"{}\", if that is a valid option you might try passing it before \"{}\" command.'
```
```suggestion
INVALID_CLI_COMMAND_ARGUMENT = 'Error: \"{}\" is not a valid cli-command argument. If \"{}\" is a valid option try passing it before the \"{}\" command.'
```

pablomartin4btc replied:
done

pablomartin4btc replied:
On this one, when I raise the error or when I have to pass the strings for the insertion into the chain, it doesn't look right to pass twice the same variable (e.g.: `INVALID_CLI_COMMAND_ARGUMENT.format(rpcwallet3, rpcwallet3, "-generate"`).

Would you accept:

`INVALID_CLI_COMMAND_ARGUMENT = 'Error: invalid cli-command argument. If \"{}\" is a valid option try passing it before the \"{}\" command.'`

andrewtoth commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -28,7 +28,10 @@
+WALLET_NOT_SPECIFIED = 'For the CLI, specify the "-rpcwallet=<filename>" option before the command'
+INVALID_CLI_COMMAND_ARGUMENT = 'Error: this is not a valid cli-command argument \"{}\", if that is a valid option you might try passing it before \"{}\" command.'
+DUPLICATE_COMMAND_ERROR = 'Error: duplicate command line argument {}'
+MULTIPLE_CLI_COMMAND_ERROR = "Error: you can run only one cli-command at a time, please choose one: {} or {}"
```
```suggestion
MULTIPLE_CLI_COMMAND_ERROR = "Error: you can only run one cli-command at a time, either {} or {}"
```

pablomartin4btc replied:
done

andrewtoth commented on src/bitcoin-cli.cpp:
```diff
@@ -887,7 +887,8 @@ static void ParseError(const UniValue& error, std::string& strPrint, int& nRet)
         if (err_code.isNum() && err_code.getInt<int>() == RPC_WALLET_NOT_SPECIFIED) {
-            strPrint += "\nTry adding \"-rpcwallet=<filename>\" option to bitcoin-cli command line.";
+            strPrint += "\nUsing \"bitcoin-cli\", add the \"-rpcwallet=<filename>\" option before the command.";
+            strPrint += "\nRun \"bitcoin-cli -h\" for help or \"bitcoin-cli listwallets\" to see which wallets are currently loaded).";
```
```suggestion
            strPrint += "\nRun \"bitcoin-cli -h\" for help or \"bitcoin-cli listwallets\" to see which wallets are currently loaded.";
```

pablomartin4btc replied:
fixed

andrewtoth commented:
ACK 0fe610deba336f0370d68c130dc9a223b7db964e modulo the dangling closing parenthesis in the error message.
Also, the first line of the commit message should be at most 80 chars.

pablomartin4btc commented:
> ACK [0fe610d](https://github.com/bitcoin/bitcoin/commit/0fe610deba336f0370d68c130dc9a223b7db964e) modulo the dangling closing parenthesis in the error message. 
> Also, the first line of the commit message should be at most 80 chars.

I've corrected them both. Thanks!

hernanmarino reviewed (approved):
ACK a870f5affcfb85f91afe989db0824313985158d8

luke-jr commented on src/bitcoin-cli.cpp:
```diff
@@ -1078,6 +1079,68 @@ static UniValue GetNewAddress()
+            // Check if the "-" command was recognised by the ArgsManager or interpreted as a parameter
+            // e.g.: bitcoin-cli -generate 1 -rpcwallet=xxx, is an invalid call
+            if (!gArgs.IsArgSet(command) && exclusivelyCliCommand != "") {
+                throw std::runtime_error(strprintf("Error: invalid cli-command argument. If \"%s\" is a valid option try passing it before the \"%s\" command.", argv[i], exclusivelyCliCommand));
```
Is there a reason we shouldn't just support this? Normally it is okay to pass switch options in any order.

pablomartin4btc replied:
Perhaps we should, this issue has been here for a while it seems (#1044). I'll ping you with an alternative I've mentioned in a [comment](https://github.com/bitcoin/bitcoin/pull/26990#issuecomment-1411046166) above (3rd paragraph).

andrewtoth commented on src/bitcoin-cli.cpp:
```diff
@@ -1101,6 +1101,68 @@ static UniValue GetNewAddress()
+    const std::set<std::string> cliCommands{"-generate", "-netinfo", "-addrinfo", "-getinfo"};
+
+    // return if element was found or not
+    return (cliCommands.find(cliCommand) != cliCommands.end());
```
Here we are using `std::set::find` and below on line 1141 we are using `std::set::count` to determine if an element already exists. We should use a consistent method for each. I'm not sure if C++20 is supported in this codebase yet, but we should use `std::set::contains` if it is.

pablomartin4btc replied:
It makes sense, I'll use `std::set::find` instead of `std::set::count` (L#1141).

Regarding C++20. it's [not enabled](https://github.com/bitcoin/bitcoin/blob/master/configure.ac#L99) yet but perhaps will be soon (#28349). If this PR doesn't get merged first I'll update the code as you suggested so I'm leaving this conversation as unresolved, otherwise I'll do a follow-up.

pablomartin4btc replied:
@andrewtoth, now I've included `std::set::contains` as [you](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1383787269) and @jonatack [recommended](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1476755003).

andrewtoth reviewed (commented):
style-nit: all variable and function parameter names should be snake_case https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c.

pablomartin4btc commented:
> style-nit: all variable and function names should be snake_case https://github.com/bitcoin/bitcoin/blob/master/doc/developer-notes.md#coding-style-c.

Thanks for checking that, you are right, I'll fix them all ASAP.

pablomartin4btc commented:
I've addressed feedback from @andrewtoth - thank you!

andrewtoth reviewed (approved):
ACK 755320f75f2141909e84b62f420462c1c5b193e6

pablomartin4btc commented on src/bitcoin-cli.cpp:
```diff
@@ -1101,6 +1101,68 @@ static UniValue GetNewAddress()
+bool IsExclusivelyCliCommand(const std::string cli_command)
+{
+    // Initialize set with cli-commands that run exclusively
+    const std::set<std::string> exclusively_commands{"-generate", "-netinfo", "-addrinfo", "-getinfo"};
```
I'd prefer to keep the definition as is since the function is called only once, thanks for the suggestion.

andrewtoth replied:
Sorry, not sure what you mean here. It looks to me like this function is called for every argument passed to bitcoin-cli, and therefore this set will be initialized for every argument instead of just once.

pablomartin4btc replied:
Right, sorry, I haven't expressed myself correctly, I meant that the validation was only "used" once but you are right, the `IsExclusivelyCliCommand` function is called for every argument passed to `bitcoin-cli`, still, `static` would keep the set  throughout the lifetime of the program.

pablomartin4btc commented on src/bitcoin-cli.cpp:
```diff
@@ -1101,6 +1101,68 @@ static UniValue GetNewAddress()
+/**
+ * ValidateCliCommand checks for:
+ *  1. duplicate command line arguments
+ *  2. no multiple bitcoin-cli side commands allowance
```
Yeah, your version is clearer. Thanks!

pablomartin4btc commented on src/bitcoin-cli.cpp:
```diff
@@ -1101,6 +1101,68 @@ static UniValue GetNewAddress()
+ * ValidateCliCommand checks for:
+ *  1. duplicate command line arguments
+ *  2. no multiple bitcoin-cli side commands allowance
+ *  3. unrecognised cli-command or any other argument starting with a slash
```
Thanks, also removed the following line and tweaked the return statement.

andrewtoth commented:
It looks safe to me to replace all uses of `std::string` with `std::string_view` in this PR.

jonatack reviewed (commented):
Reviewed the first commit fa48d460334beef43fa73455aa9e30971d33c1b2. It looks like the prefix in its commit name should be `cli` instead of `gui`. Will review the second commit soon.

<details><summary>combined suggested diff for <code>fa48d46</code></summary><p>

```diff
diff --git a/src/bitcoin-cli.cpp b/src/bitcoin-cli.cpp
index a53c2a1b0d0..b342382fc96 100644
--- a/src/bitcoin-cli.cpp
+++ b/src/bitcoin-cli.cpp
@@ -53,6 +53,7 @@ using CliClock = std::chrono::system_clock;
 const std::function<std::string(const char*)> G_TRANSLATION_FUN = nullptr;
 UrlDecodeFn* const URL_DECODE = urlDecode;
 
+constexpr std::array<std::string_view, 4> CLI_COMMANDS{"-netinfo", "-getinfo", "-addrinfo", "-generate"};
 static const char DEFAULT_RPCCONNECT[] = "127.0.0.1";
 static const int DEFAULT_HTTP_CLIENT_TIMEOUT=900;
 static constexpr int DEFAULT_WAIT_CLIENT_TIMEOUT = 0;
@@ -1105,20 +1106,6 @@ static UniValue GetNewAddress()
     return ConnectAndCallRPC(&rh, "getnewaddress", /* args=*/{}, wallet_name);
 }
 
-/**
- * IsExclusivelyCliCommand checks if a string is a cli-command that needs to run exclusively
- * @param[in] cli_command The string to check if it is a cli-command that needs to run exclusively.
- * @returns true if cli_command is included in the set of the exclusively_commands.
- */
-bool IsExclusivelyCliCommand(const std::string_view cli_command)
-{
-    // Initialize set with cli-commands that run exclusively
-    const std::set<std::string_view> exclusively_commands{"-generate", "-netinfo", "-addrinfo", "-getinfo"};
-
-    // return if element was found or not
-    return (exclusively_commands.find(cli_command) != exclusively_commands.end());
-}
-
 /**
  * ValidateCliCommand checks for:
  *  1. duplicate command line arguments
@@ -1131,38 +1118,26 @@ static void ValidateCliCommand(int argc, char *argv[])
     std::set<std::string_view> commands;
     std::string_view exclusively_command;
     for (int i = 1; i < argc; ++i) {
-        if (IsSwitchChar(argv[i][0])) {
-            std::string_view command(argv[i]);
-            // Check if the command line argument has an =, in that case, ignore what's after;
-            // e.g.: -color=never; we just want to check "-color"
-            std::size_t pos = command.find('=');
-            if (pos != std::string::npos) {
-                command = command.substr(0, pos);
-            }
-
-            // Check if this command has already been seen
-            if (commands.find(command) != commands.end()) {
-                throw std::runtime_error(strprintf("Error: duplicate command line argument %s", command));
-            }
-
-            // Check if it's a cli-command that runs exclusively e.g. a user could run:
-            // "bitcoin-cli -regtest -generate -netinfo -addrinfo -getinfo" but it's confusing,
-            // only the last one will be executed so we make sure the user specify only one
-            if (IsExclusivelyCliCommand(command)) {
-                if (!exclusively_command.empty()) {
-                    throw std::runtime_error(strprintf("Error: you can only run one cli-command at a time, either %s or %s", exclusively_command, command));
-                }
-                exclusively_command = command;
-            }
-
-            // Check if the "-" command was recognised by the ArgsManager or interpreted as a parameter
-            // e.g.: bitcoin-cli -generate 1 -rpcwallet=xxx, is an invalid call
-            if (!gArgs.IsArgSet(std::string(command)) && !exclusively_command.empty()) {
-                throw std::runtime_error(strprintf("Error: invalid cli-command argument. If \"%s\" is a valid option try passing it before the \"%s\" command.", argv[i], exclusively_command));
+        if (!IsSwitchChar(argv[i][0])) continue;
+        std::string_view command(argv[i]);
+        if (const auto pos{command.find_first_of('=')}; pos != std::string_view::npos) {
+            command = command.substr(0, pos);
+        }
+        if (commands.contains(command)) {
+            throw std::runtime_error(strprintf("duplicate bitcoin-cli command %s", command));
+        }
+        if (std::find(std::begin(CLI_COMMANDS), std::end(CLI_COMMANDS), command) != std::end(CLI_COMMANDS)) {
+            if (!exclusively_command.empty()) {
+                throw std::runtime_error(strprintf("you can only run one bitcoin-cli command at a time, either %s or %s", exclusively_command, command));
             }
-
-            commands.insert(command);
+            exclusively_command = command;
+        }
+        // Check if the "-" command was recognised by the ArgsManager or interpreted as a parameter
+        // e.g.: bitcoin-cli -generate 1 -rpcwallet=xxx, is an invalid call
+        if (!exclusively_command.empty() && !gArgs.IsArgSet(std::string(command))) {
+            throw std::runtime_error(strprintf("invalid bitcoin-cli argument; if \"%s\" is a valid option, try passing it before the \"%s\" command", argv[i], exclusively_command));
         }
+        commands.insert(command);
     }
 }
 
diff --git a/test/functional/interface_bitcoin_cli.py b/test/functional/interface_bitcoin_cli.py
index 2a64c78f216..057e4a2f389 100755
--- a/test/functional/interface_bitcoin_cli.py
+++ b/test/functional/interface_bitcoin_cli.py
@@ -29,9 +29,9 @@ BLOCKS_VALUE_OF_ZERO = 'error: the first argument (number of blocks to generate,
 TOO_MANY_ARGS = 'error: too many arguments (maximum 2 for nblocks and maxtries)'
 WALLET_NOT_LOADED = 'Requested wallet does not exist or is not loaded'
 WALLET_NOT_SPECIFIED = "Multiple wallets are loaded. Please specify a wallet to use"
-INVALID_CLI_COMMAND_ARGUMENT = 'Error: invalid cli-command argument. If \"{}\" is a valid option try passing it before the \"{}\" command.'
-DUPLICATE_COMMAND_ERROR = 'Error: duplicate command line argument {}'
-MULTIPLE_CLI_COMMAND_ERROR = "Error: you can only run one cli-command at a time, either {} or {}"
+INVALID_CLI_COMMAND_ARGUMENT = 'error: invalid bitcoin-cli argument; if \"{}\" is a valid option, try passing it before the \"{}\" command'
+DUPLICATE_COMMAND_ERROR = 'error: duplicate bitcoin-cli command {}'
+MULTIPLE_CLI_COMMAND_ERROR = "error: you can only run one bitcoin-cli command at a time, either {} or {}"
 
 
 def cli_get_info_string_to_dict(cli_get_info_string):
@@ -327,7 +327,7 @@ class TestBitcoinCli(BitcoinTestFramework):
             self.log.info('Test -generate with bad args & without -rpcwallet in multiwallet mode')
             assert_raises_rpc_error(-19, WALLET_NOT_SPECIFIED, self.nodes[0].cli('-generate', 1, 2, 3).echo)
 
-            self.log.info('Test running multiple cli-commands in one line')
+            self.log.info('Test running multiple CLI commands in one line')
```
</p></details>

jonatack reviewed (commented):
Approach ACK modulo the review suggestions.

hernanmarino reviewed (commented):
re ACK fa48d460334beef43fa73455aa9e30971d33c1b2
Just a nit (only if you update) : your second commit title seems to suggest that it's gui related, instead of cli.
Also, and although not invalidating for me , I think that @jonatack 's suggestions are worth following.

pablomartin4btc commented:
> Are you still working on this?

I'm about to continue, sorry for the delay!

pablomartin4btc commented:
Updates:
- I've been working on @jonatack's latest feedback, I'm performing some tests at the moment and will update the code soon.

pablomartin4btc commented on src/bitcoin-cli.cpp:
```diff
@@ -908,7 +908,8 @@ static void ParseError(const UniValue& error, std::string& strPrint, int& nRet)
         }
         if (err_code.isNum() && err_code.getInt<int>() == RPC_WALLET_NOT_SPECIFIED) {
-            strPrint += "\nTry adding \"-rpcwallet=<filename>\" option to bitcoin-cli command line.";
+            strPrint += "\nUsing \"bitcoin-cli\", add the \"-rpcwallet=<filename>\" option before the command.";
```
I've made a combination of current state and your improvements, based on the feedback provided by @andrewtoth ([1](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1127232884), [2](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1127235671)). Please let me know what you think.

pablomartin4btc replied:
Ok, I'll work on the full error message verification in the test and will consider your preference on the error documentation. Thanks again!

pablomartin4btc replied:
Ok, thanks for checking, I'll work on it and will incorporate full error message in the functional test.

pablomartin4btc replied:
Done.

pablomartin4btc commented on src/bitcoin-cli.cpp:
```diff
@@ -1101,6 +1101,67 @@ static UniValue GetNewAddress()
+
+            // Check if this command has already been seen
+            if (commands.find(command) != commands.end()) {
+                throw std::runtime_error(strprintf("Error: duplicate command line argument %s", command));
```
It makes sense. Thanks!

pablomartin4btc commented on src/bitcoin-cli.cpp:
```diff
@@ -1101,6 +1101,67 @@ static UniValue GetNewAddress()
+            // only the last one will be executed so we make sure the user specify only one
+            if (IsExclusivelyCliCommand(command)) {
+                if (!exclusively_command.empty()) {
+                    throw std::runtime_error(strprintf("Error: you can only run one cli-command at a time, either %s or %s", exclusively_command, command));
```
[Same](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1573464459). Done!

pablomartin4btc commented on src/bitcoin-cli.cpp:
```diff
@@ -1101,6 +1101,67 @@ static UniValue GetNewAddress()
+            // Check if the "-" command was recognised by the ArgsManager or interpreted as a parameter
+            // e.g.: bitcoin-cli -generate 1 -rpcwallet=xxx, is an invalid call
+            if (!gArgs.IsArgSet(std::string(command)) && !exclusively_command.empty()) {
+                throw std::runtime_error(strprintf("Error: invalid cli-command argument. If \"%s\" is a valid option try passing it before the \"%s\" command.", argv[i], exclusively_command));
```
[Same](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1573464459). Done!

pablomartin4btc commented on src/bitcoin-cli.cpp:
```diff
@@ -1101,6 +1101,67 @@ static UniValue GetNewAddress()
+            }
+
+            // Check if this command has already been seen
+            if (commands.find(command) != commands.end()) {
```
It was also [suggested](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1383787269) by @andrewtoth, but C++20 wasn't supported at that time. Done!

pablomartin4btc commented on src/bitcoin-cli.cpp:
```diff
@@ -1101,6 +1101,67 @@ static UniValue GetNewAddress()
+
+    // return if element was found or not
+    return (exclusively_commands.find(cli_command) != exclusively_commands.end());
+}
```
It makes more sense, thanks! Done!

pablomartin4btc commented on src/bitcoin-cli.cpp:
```diff
@@ -1101,6 +1101,67 @@ static UniValue GetNewAddress()
+
+            // Check if the "-" command was recognised by the ArgsManager or interpreted as a parameter
+            // e.g.: bitcoin-cli -generate 1 -rpcwallet=xxx, is an invalid call
+            if (!gArgs.IsArgSet(std::string(command)) && !exclusively_command.empty()) {
```
True. Done!

pablomartin4btc commented on src/bitcoin-cli.cpp:
```diff
@@ -1101,6 +1101,67 @@ static UniValue GetNewAddress()
+            // Check if the command line argument has an =, in that case, ignore what's after;
+            // e.g.: -color=never; we just want to check "-color"
+            std::size_t pos = command.find('=');
+            if (pos != std::string::npos) {
```
I like it. Done!

pablomartin4btc commented on src/bitcoin-cli.cpp:
```diff
@@ -1101,6 +1101,67 @@ static UniValue GetNewAddress()
+    std::set<std::string_view> commands;
+    std::string_view exclusively_command;
+    for (int i = 1; i < argc; ++i) {
+        if (IsSwitchChar(argv[i][0])) {
```
Nice. Done!

pablomartin4btc commented:
Updates:
- Addressed [feedback](https://github.com/bitcoin/bitcoin/pull/26990#pullrequestreview-1860423961) provided by @jonatack regarding first commit 6898219043f42ae42e4703c549e656f56d276d13.
- Updated second commit cf7dd3564a3ace4153a32930f36bd78432b59097 following some directions [given](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1483546926) by @jonatack and [discussed](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1573464319) with @andrewtoth.
  - Fixed incorrect commit title prefix (~~`gui:`~~ => `cli:`).

jonatack reviewed (commented):
Approach ACK, modulo https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1614026697 and the other comment.

Also, "protocol.h" in the first commit message body seems wrong or out-of-date here: `bringing the error message in line with the definition established in protocol.h ("error when there are multiple wallets loaded")`. I'm not sure, but did you mean "src/wallet/rpc/util.cpp" instead?

Edit: Apologies for how long I took to re-review, will react quickly on this next iteration :)

pablomartin4btc commented on src/bitcoin-cli.cpp:
```diff
@@ -1107,6 +1111,48 @@ static UniValue GetNewAddress()
+
+        // Check if this command has already been seen
+        if (commands.contains(command)) {
+            throw std::runtime_error(strprintf("duplicate bitcoin-cli command %s.", command));
```
Ok, thanks, will do.

pablomartin4btc replied:
Done.

pablomartin4btc commented:
> Also, "protocol.h" in the first commit message body seems wrong or out-of-date here: `bringing the error message in line with the definition established in protocol.h ("error when there are multiple wallets loaded")`. I'm not sure, but did you mean "src/wallet/rpc/util.cpp" instead?

The reference to "`src/rpc/protocol.h`" was intended to clarify the context of the error message alignment, specifically in the section of the code where the RPC error code "not wallet specified" is defined (`enum RPCErrorCode`) as related to the issue of multiple wallets being loaded. Here’s the relevant code snippet for reference:

https://github.com/bitcoin/bitcoin/blob/fe70be537783aeadf9e3f72cad07efd66775285b/src/rpc/protocol.h#L82

However, if it adds more confusion than clarity, I can certainly remove it.

pablomartin4btc commented:
Updates:
- Addressed latest @jonatack's [feedback](https://github.com/bitcoin/bitcoin/pull/26990#pullrequestreview-2078072832):
  - Removed [unnecessary](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1614031474) trailing `.` from CLI error messages.
  - Reconsidered previous [suggestion](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1614026697) about the `RPC_WALLET_NOT_SPECIFIED` error handled in `src/bitcoin.cli` and the returned message.
  - Added [full error message](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1127282385) into the functional test.

pablomartin4btc commented:
Updates:
- Addressed latest @jonatack's [feedback](https://github.com/bitcoin/bitcoin/pull/26990#pullrequestreview-2078072832):
  - Removed [unnecessary](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1614031474) trailing `.` from CLI error messages.
  - Reconsidered previous [suggestion](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1614026697) about the `RPC_WALLET_NOT_SPECIFIED` error handled in `src/bitcoin.cli` and the returned message.
  - Added [full error message](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1127282385) into the functional test.

jonatack reviewed (commented):
Thanks for updating!  I think 6ccf7cb946bd3772b5fba92246be7e2281f59d09 is good to go.

WIP review on commit 3d63fc976d616436d64335b15a918ffba1883b9a but I initially find it confusing. The diffs below are between your commit and testing some modifications to it.

For instance, the tests that pass a wallet CLI option to a non-wallet CLI command (-generate) seem to be mixing issues under test...

```diff
-            # Single or Multi wallet modes don't matter here as -generate command validation is performed before we send to RPC
-            # Note also that -rpcwallet arg intentionally adds some confusion but could be any arg starting with "-"
-            self.log.info('Test -generate with nblocks and -rpcwallet afterwards')
-            assert_raises_process_error(1, INVALID_CLI_COMMAND_ARGUMENT.format(rpcwallet3, "-generate"), self.nodes[0].cli('-generate', 1, rpcwallet3).send_cli)
+            self.log.info('Test -generate followed by a CLI option (invalid order of arguments)')
+            assert_raises_process_error(1, INVALID_CLI_COMMAND_ARGUMENT.format("-rpcwait", "-generate"), self.nodes[0].cli('-generate', 4, 1000, "-rpcwait").send_cli)
```

...and at first glance I'm not sure if this is the correct behavior, but I need to look at it more

```diff 
-            # Note, same as above, -rpcwallet arg intentionally adds some confusion but could be any arg starting with "-"
-            self.log.info('Test -generate -rpcwallet with nblocks and -rpcwallet again afterwards')
-            assert_raises_process_error(1, DUPLICATE_COMMAND_ERROR.format("-rpcwallet"), self.nodes[0].cli(rpcwallet3, '-generate', 1, rpcwallet3).send_cli)
+            self.log.info('Test -generate with a duplicate CLI command')
+            assert_raises_process_error(1, DUPLICATE_COMMAND_ERROR.format("-rpcwait"), self.nodes[0].cli('-rpcwait', '-generate', 4, 1000, "-rpcwait").send_cli)
```

Will finishing reviewing that commit.

pablomartin4btc commented:
> For instance, the tests that pass a wallet CLI option to a non-wallet CLI command (-generate) seem to be mixing issues under test...
> 
> ```diff
> -            # Single or Multi wallet modes don't matter here as -generate command validation is performed before we send to RPC
> -            # Note also that -rpcwallet arg intentionally adds some confusion but could be any arg starting with "-"
> -            self.log.info('Test -generate with nblocks and -rpcwallet afterwards')
> -            assert_raises_process_error(1, INVALID_CLI_COMMAND_ARGUMENT.format(rpcwallet3, "-generate"), self.nodes[0].cli('-generate', 1, rpcwallet3).send_cli)
> +            self.log.info('Test -generate followed by a CLI option (invalid order of arguments)')
> +            assert_raises_process_error(1, INVALID_CLI_COMMAND_ARGUMENT.format("-rpcwait", "-generate"), self.nodes[0].cli('-generate', 4, 1000, "-rpcwait").send_cli)
> ```

Thanks for double checking the tests. For the one above, the idea was to test the proper use case of the issue described at the very top of the PR's desciption, would fail with any "-" option, as explained, I've also clarified that could add confusion but it was intentional. Again, if you think it adds more confusion, I'm happy to remove it (or keep it) and add your suggestion anyways.

> ...and at first glance I'm not sure if this is the correct behavior, but I need to look at it more
> 
> ```diff
> -            # Note, same as above, -rpcwallet arg intentionally adds some confusion but could be any arg starting with "-"
> -            self.log.info('Test -generate -rpcwallet with nblocks and -rpcwallet again afterwards')
> -            assert_raises_process_error(1, DUPLICATE_COMMAND_ERROR.format("-rpcwallet"), self.nodes[0].cli(rpcwallet3, '-generate', 1, rpcwallet3).send_cli)
> +            self.log.info('Test -generate with a duplicate CLI command')
> +            assert_raises_process_error(1, DUPLICATE_COMMAND_ERROR.format("-rpcwait"), self.nodes[0].cli('-rpcwait', '-generate', 4, 1000, "-rpcwait").send_cli)
> ```

In this case, similar to the previous, without taking into account the order, wanted to use `-rpcwallet` option but could take (replace/ add?) the `-rpcwait` case as you are suggesting.

pablomartin4btc commented:
> On my short list to re-review.

Thanks, need to rebase but first I have to remove the validation for the CLI commands that seems it has been improved in #30148 which I missed to review.

pablomartin4btc commented:
> (I think this could be merged much more quickly if it only contained [6ccf7cb](https://github.com/bitcoin/bitcoin/commit/6ccf7cb946bd3772b5fba92246be7e2281f59d09) and proposed the validation in a separate pull.)

I do agree and thought many times about it haha... I'll split it out tomorrow. Thanks.

mzumsande commented on src/bitcoin-cli.cpp:
```diff
@@ -947,7 +947,8 @@ static void ParseError(const UniValue& error, std::string& strPrint, int& nRet)
         }
         if (err_code.isNum() && err_code.getInt<int>() == RPC_WALLET_NOT_SPECIFIED) {
-            strPrint += "\nTry adding \"-rpcwallet=<filename>\" option to bitcoin-cli command line.";
+            strPrint += " Or for the CLI, specify the \"-rpcwallet=<filename>\" option before the command";
```
I found "filename" confusing, see #30912 - if there isn't a deeper reason behind it that I'm unaware of, it could be renamed to `walletname`.

furszy replied:
I think it would be beneficial to explain the alternatives. The `-rpcwallet` parameter depends on how the wallet is loaded. It can either be the wallet directory name located inside the `-walletdir` (the default value is `datadir/net/wallets/wallet_name`, in which case only `wallet_name` is used in this option), or the absolute path to the wallet directory, which is the complete path `datadir/net/wallets/wallet_name`.

You can note the difference by loading an existing wallet in two different ways:
1) `loadwallet "test_wallet"`
2) `loadwallet "datadir/regtest/wallets/test_wallet"`

Then, call `getwalletinfo()`. The wallet name will differ, and as a result, the `-rpcwallet=<arg>` argument for the very same wallet will be different as well.

mzumsande replied:
... and I think in both of these cases `walletname` would be correct and `filename` incorrect, it would just be a different wallet name depending on how it's loaded.

It could actually also be a filename: Put a legacy wallet `oldwallet.dat` into the datadir, run with `-deprecatedrpc=create_bdb` and `loadwallet "oldwallet.dat"` will load it successfully, in this special case the filename is the walletname. But even here, where `filename` would make sense, `walletname` is correct too, so it just seem better overall.

pablomartin4btc commented:
Updates:
- Removed validation commit which I'll added in a separate PR as [suggested](https://github.com/bitcoin/bitcoin/pull/26990#issuecomment-2354091658) by @jonatack.
- Updated description and title accordingly with above. 
- Added test checking -rpcwallet=<filename> is not doable as [suggested](https://github.com/bitcoin/bitcoin/pull/26990#discussion_r1762109275) by @jonatack.
- Replaced `-rpcwallet=<filename>` which is incorrect with `-rpcwallet=<walletname>` as suggested by @mzumsande.

furszy reviewed (commented):
utACK 54227e681a4

mzumsande reviewed (commented):
Code Review ACK 54227e681a4efa8961f1ad05d43366d88a9b686a

jonatack reviewed (commented):
ACK 54227e681a4efa8961f1ad05d43366d88a9b686a

pablomartin4btc commented on test/functional/interface_bitcoin_cli.py:
```diff
@@ -331,6 +336,10 @@ def run_test(self):
             n4 = 10
             blocks = self.nodes[0].getblockcount()
 
+            self.log.info('Test -generate -rpcwallet=<filename> raise RPC error')
```
I'll do this on the follow-up where I'll be adding the validation from the 2nd commit that was dropped from this PR.

achow101 commented:
ACK 54227e681a4efa8961f1ad05d43366d88a9b686a

luke-jr commented:
> Removed validation commit which I'll added in a separate PR as https://github.com/bitcoin/bitcoin/pull/26990#issuecomment-2354091658 by @jonatack.

Does this exist?

pablomartin4btc commented:
> > Removed validation commit which I'll added in a separate PR as [#26990 (comment)](https://github.com/bitcoin/bitcoin/pull/26990#issuecomment-2354091658) by @jonatack.
> 
> Does this exist?

So, this was the 2nd commit https://github.com/bitcoin/bitcoin/commit/3d63fc976d616436d64335b15a918ffba1883b9a that I've removed for simplicity of this PR plus one of those validations: "only 1 cli-command can run at a time (eg can't run -generate and
  -getinfo at the same time)," was done nicely on a separate PR that I missed reviewing (#30148).

The other 2 validations on that commit were:
- (a) no params starting with slash "-" will be accepted after a cli-command (this is the case for bitcoin-cli -generate 3 -rpcwallet=xyz).
- (b) duplication of cli-command (and options: at the moment a user can specify -rpcwallet many times, only the last one will be
  taken into account.
  
After speaking to @jonatack offline, (b) is something that doesn't fail, the command still does what is intended to do, but (a), apart from the `generate` command that probably not used as much, when you run other commands like `./build/src/bitcoin-cli -rpcwallet=abc getaddressinfo <address> -rpcwallet=abc` will return the RPC help and this `./build/src/bitcoin-cli -regtest listtransactions -rpcwallet=t3` won't return anything, so as also @kouloumos [mentioned](https://github.com/bitcoin/bitcoin/pull/26990#pullrequestreview-1274911604) earlier, the order of the [options] matter and creates confusion at the time of evaluate the arguments of a RPC command. Because of that, instead of adding restrictions in `bitcoin-cli.cpp` I thought it was better to identify those [options] in the `ArgsMan` while making the code cleaner in `bitcoin-cli.cpp`.

I've made that work in 416cbb9aa434560898ba1e5c3724ca69e3f5994c, which needs a bit of cleanup (just left some commented lines if we wanted to update the value of the [option] instead of adding infinite values as the current behaviour but that change will break some tests) and I think I sill have to verify some tests and add new ones if needed. If this logic is something that sounds correct to you and you are interested I can open a PR soon in a couple of days, please let me know! Thank you!
